### PR TITLE
Tnr add row label options

### PR DESCRIPTION
**Pull request by ericsturman**

### CODE REVIEW

**Comment by tnrich** [comment]
Nice work @ericsturman, this is looking good! There are a few small things that I think we'll want to improve on before this gets merged. 

- [ ] make lines extend all the way to the center of the annotation 
- [ ] give the lines a default opacity of .3 or something in that ballpark so that reading other labels is easier
- [ ] add a way to "relax" the labels so that more can fit on a single line (might be difficult)
- [ ] maybe add this as a user configurable option from the File > View menu (thoughts?)

I also have a couple questions. You said: 

>@tnrich I have pushed out changes to the branch you created. Things mostly went as expected except I decided to move parts down under the labels when part/feature labels are external so that the labels point down toward the all PointedAnnotations. I also needed to filter out certain annotation ranges (line 44 in labels.js); I'm not sure if there was something I was not understanding about annotations vs annotationRanges or not. I also noticed that some labels are colliding in the Linear Map and it's not immediately obvious to me how to fix this.

What do you mean exactly by "some labels are colliding"? Like they're getting drawn over eachother? Or that they're not getting drawn at all? 


Also, did you have a more specific question about: 
>I'm not sure if there was something I was not understanding about annotations vs annotationRanges or not.

File: src/RowItem/Labels.js [hunk 1]
```diff
@@ -37,6 +40,27 @@ function Labels(props) {
   let annotationsSVG = [];
   let rowCenter = rowLength / 2;
   let iTree = new IntervalTree(rowCenter);
+
+  annotationRanges = values(
+    reduce(
+      annotationRanges,
+      (accum, annotationRange) => {
+        if (
+          annotationRange.annotation.annotationTypePlural === "parts" ||
```
Comment (tnrich) [inline src/RowItem/Labels.js]: Why are parts being let through here but other annotations not?

File: src/RowItem/Labels.js [hunk 2]
```diff
@@ -37,6 +40,27 @@ function Labels(props) {
...
+          !accum[annotationRange.id] ||
```
Comment (tnrich) [inline src/RowItem/Labels.js]: We'll want to prefix the id with the annotationTypePlural so it looks like 
```
!accum[
            annotationRange.annotation.annotationTypePlural +
              "--" +
              annotationRange.id
          ]
``` 

kind of verbose but necessary if id's are re-used across annotation types

**Comment by tnrich** [comment]
@ericsturman I'm experimenting a little bit with the label drawing. What do you think about having the external labels being split onto different levels to be closer to the annotation type of interest? :  
image

**Comment by tnrich** [comment]
@ericsturman also, what do you think about having an option like "externalLabelsOnlyIfNecessary"? 

The idea being that any annotation that can fit the label within itself wouldn't get an external label.

**Comment by ericsturman** [comment]
@tnrich I like all of the ideas in the first checklist.  Also, "colliding labels" means they are getting drawn over each other.  In terms of maybe not understanding annotation ranges, it seemed that in the case of the araC feature I was getting  4 annotation range items:  3 for each of the sub-locations and then one for the full length feature...that's why I wrote that reduce code...and I believe that the parts did not have an ID IIRC so I had to exclude them from the filtering step and assume they would never look like araC.  I think the labels separated by type would be easier to read and I definitely like the "only if necessary" option.

**Comment by tnrich** [comment]
@ericsturman what's this piece of code doing: 
```js 
accum[identifier].end - accum[identifier].start <
            annotationRange.end - annotationRange.start
```
that's basically checking for the "largest" chunk, right? If so I'll have it use the `getRangeLength()` fn instead

**Comment by ericsturman** [comment]
That's exactly it... `getRangeLength()` would be better...

**Comment by tnrich** [comment]
@ericsturman I pushed my changes to https://github.com/TeselaGen/openVectorEditor/pull/581 
I fixed a bunch of miscellaneous stuff plus the the things I checked off above: 
- [x] make lines extend all the way to the center of the annotation when hovered
- [x]  give the lines a default opacity of .3 or something in that ballpark so that reading other labels is easier
- [x] add primers to the list of things that can be drawn with external labels

 If you want to fork off of that and finish some of the changes I was working on that would be great! Specifically I think we still need to: 
- [ ] finish implementing the `onlyShowLabelsThatDoNotFit` option (I think that should be the default when external labels is enabled) 
- [ ] change cutsites to be italicized in circular view to better match row view
- [ ] add this as a user configurable option from a new File > View > External Labels menu and have the user selection persisted to localstorage by default. I can help more with the implementation details on this one 

Let me know if you think you can get those done. We're getting close :)

**Comment by tnrich** [comment]
@ericsturman I pushed my changes to https://github.com/TeselaGen/openVectorEditor/pull/581 
I fixed a bunch of miscellaneous stuff plus the the things I checked off above: 
- [x] make lines extend all the way to the center of the annotation when hovered
- [x]  give the lines a default opacity of .3 or something in that ballpark so that reading other labels is easier
- [x] add primers to the list of things that can be drawn with external labels

 If you want to fork off of that and finish some of the changes I was working on that would be great! Specifically I think we still need to: 
- [ ] finish implementing the `onlyShowLabelsThatDoNotFit` option (I think that should be the default when external labels is enabled) 
- [ ] change cutsites to be italicized in circular view to better match row view
- [ ] add this as a user configurable option from a new File > View > External Labels menu and have the user selection persisted to localstorage by default. I can help more with the implementation details on this one 

Let me know if you think you can get those done. We're getting close :)

**Comment by ericsturman** [comment]
@tnrich I will get started on the first two and then we can chat about the third.

**Comment by ericsturman** [comment]
@tnrich We have a busy couple of days... I have made the changes and sent out some screenshots but we still need to discuss as a team... I will be getting back to you soon... hopefully with final code changes...

**Pull request closed without merging**